### ISSUE TITLE
ONTOP multi line optional Block Error

### ISSUE THREAD

**Issue by carstendraschner** [SANSA-Stack/SANSA-Stack#122]
When I try to run a sparql query with multi line optional blocks I get error:
```
20/12/15 11:16:44 ERROR OntopSPARQLEngine: Spark failed to execute translated SQL query


SELECT ?seed ?seed__down_hasSpouse__down_name
WHERE {
?seed a <http://dig.isi.edu/Person> .
OPTIONAL {
 ?seed <http://dig.isi.edu/hasSpouse> ?seed__down_hasSpouse .
 ?seed__down_hasSpouse <http://dig.isi.edu/name> ?seed__down_hasSpouse__down_name .
}
}
org.apache.spark.sql.catalyst.parser.ParseException: 
mismatched input 'ON' expecting {')', '.', '[', 'AND', 'BETWEEN', 'DIV', 'IN', 'IS', 'LIKE', NOT, 'OR', RLIKE, EQ, '<=>', '<>', '!=', '<', LTE, '>', GTE, '+', '-', '*', '/', '%', '&', '|', '||', '^'}(line 7, pos 99)

== SQL ==
SELECT V5.`o1m5` AS `o1m5`, V5.`s1m1` AS `s1m1`
FROM (SELECT DISTINCT V2.`o` AS `o1m2`, V3.`o` AS `o1m5`, V1.`s` AS `s1m1`, CASE WHEN V2.`o` IS NOT NULL THEN V1.`s` ELSE NULL END AS `s1m2`
FROM `httpP3aP2fP2fwwwCw3CorgP2f1999P2f02P2f22dashrdfdashsyntaxdashnsP23type` V1
 LEFT OUTER JOIN 
`httpP3aP2fP2fdigCisiCeduP2fhasspouse` V2
 JOIN 
`httpP3aP2fP2fdigCisiCeduP2fname_xmlschemaP23string_lang` V3 ON (V2.`o` = V3.`s` AND '' = V3.`l`)  ON V1.`s` = V2.`s` 
---------------------------------------------------------------------------------------------------^^^
WHERE 'http://dig.isi.edu/Person' = V1.`o`
) V5

	at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:266)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:133)
	at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:48)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:81)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:605)
	at org.apache.spark.sql.SparkSession$$Lambda$2272/0x0000000050f0e620.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:605)
	at org.apache.spark.sql.SparkSession$$Lambda$2271/0x0000000050f0e020.apply(Unknown Source)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:764)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:602)
	at net.sansa_stack.query.spark.ontop.OntopSPARQLEngine.executeDebug(OntopSPARQLEngine.scala:328)
	at net.sansa_stack.query.spark.ontop.OntopSPARQLEngine.execSelect(OntopSPARQLEngine.scala:372)
	at net.sansa_stack.ml.spark.featureExtraction.SparqlFrame.getResultBindings(SparqlFrame.scala:47)
	at net.sansa_stack.ml.spark.featureExtraction.SparqlFrame.transform(SparqlFrame.scala:55)
	at net.sansa_stack.ml.spark.featureExtraction.SampleFeatureExtractionPipeline$.main(SampleFeatureExtractionPipeline.scala:58)
	at net.sansa_stack.ml.spark.featureExtraction.SampleFeatureExtractionPipeline.main(SampleFeatureExtractionPipeline.scala)
Exception in thread "main" org.apache.spark.sql.catalyst.parser.ParseException: 
mismatched input 'ON' expecting {')', '.', '[', 'AND', 'BETWEEN', 'DIV', 'IN', 'IS', 'LIKE', NOT, 'OR', RLIKE, EQ, '<=>', '<>', '!=', '<', LTE, '>', GTE, '+', '-', '*', '/', '%', '&', '|', '||', '^'}(line 7, pos 99)

== SQL ==
SELECT V5.`o1m5` AS `o1m5`, V5.`s1m1` AS `s1m1`
FROM (SELECT DISTINCT V2.`o` AS `o1m2`, V3.`o` AS `o1m5`, V1.`s` AS `s1m1`, CASE WHEN V2.`o` IS NOT NULL THEN V1.`s` ELSE NULL END AS `s1m2`
FROM `httpP3aP2fP2fwwwCw3CorgP2f1999P2f02P2f22dashrdfdashsyntaxdashnsP23type` V1
 LEFT OUTER JOIN 
`httpP3aP2fP2fdigCisiCeduP2fhasspouse` V2
 JOIN 
`httpP3aP2fP2fdigCisiCeduP2fname_xmlschemaP23string_lang` V3 ON (V2.`o` = V3.`s` AND '' = V3.`l`)  ON V1.`s` = V2.`s` 
---------------------------------------------------------------------------------------------------^^^
WHERE 'http://dig.isi.edu/Person' = V1.`o`
) V5

	at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:266)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:133)
	at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:48)
	at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:81)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:605)
	at org.apache.spark.sql.SparkSession$$Lambda$2272/0x0000000050f0e620.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:605)
	at org.apache.spark.sql.SparkSession$$Lambda$2271/0x0000000050f0e020.apply(Unknown Source)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:764)
	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:602)
	at net.sansa_stack.query.spark.ontop.OntopSPARQLEngine.executeDebug(OntopSPARQLEngine.scala:328)
	at net.sansa_stack.query.spark.ontop.OntopSPARQLEngine.execSelect(OntopSPARQLEngine.scala:372)
	at net.sansa_stack.ml.spark.featureExtraction.SparqlFrame.getResultBindings(SparqlFrame.scala:47)
	at net.sansa_stack.ml.spark.featureExtraction.SparqlFrame.transform(SparqlFrame.scala:55)
	at net.sansa_stack.ml.spark.featureExtraction.SampleFeatureExtractionPipeline$.main(SampleFeatureExtractionPipeline.scala:58)
	at net.sansa_stack.ml.spark.featureExtraction.SampleFeatureExtractionPipeline.main(SampleFeatureExtractionPipeline.scala)
```

I also tried this example without optional block like: 
```
SELECT ?seed ?seed__down_hasSpouse__down_name
WHERE {
?seed a <http://dig.isi.edu/Person> .
?seed <http://dig.isi.edu/hasSpouse> ?seed__down_hasSpouse .
?seed__down_hasSpouse <http://dig.isi.edu/name> ?seed__down_hasSpouse__down_name .
}
```
and this worked but for sure it would be desireable to have also optional blocks.
with sparqlify the multiline optional blocks run trough. the only opportunity there is that currently I am missing a easy way to get only columns specified in select. so what is the shortest postprocessing to get it
```
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
20/12/15 11:39:34 INFO RdfPartitionUtilsSpark$: started vertical partitioning of the data ...
20/12/15 11:39:35 WARN AbstractLineBasedNodeTupleReader: Configured to ignore bad tuples, parsing errors will be logged and further parsing aborted but no user visible errors will be thrown.  Consider setting rdf.io.input.ignore-bad-tuples to false to disable this behaviour
20/12/15 11:39:35 INFO AbstractLineBasedNodeTupleReader: Got split with start 0 and length 568 for file with total length of 568
20/12/15 11:39:35 INFO RdfPartitionUtilsSpark$: ... finished vertical partitioning of the data.
20/12/15 11:39:35 WARN TypeSystemImpl: Skipping: date, date
20/12/15 11:39:35 WARN TypeSystemImpl: Skipping: integer, integer
20/12/15 11:39:35 WARN TypeSystemImpl: Skipping: float, float
20/12/15 11:39:35 WARN TypeSystemImpl: Skipping: geography, geography
20/12/15 11:39:35 WARN TypeSystemImpl: Skipping: geometry, geometry
20/12/15 11:39:37 WARN TypeSystemImpl: Skipping: date, date
20/12/15 11:39:37 WARN TypeSystemImpl: Skipping: integer, integer
20/12/15 11:39:37 WARN TypeSystemImpl: Skipping: float, float
20/12/15 11:39:37 WARN TypeSystemImpl: Skipping: geography, geography
20/12/15 11:39:37 WARN TypeSystemImpl: Skipping: geometry, geometry
20/12/15 11:39:37 INFO SchemaProvider: Retrieving schema for query: SELECT * FROM `http://dig.isi.edu/name_XMLSchema#string_lang`
20/12/15 11:39:37 WARN SchemaProvider: Using ugly hack for adding a limit
20/12/15 11:39:38 INFO SchemaProvider: s -> string
20/12/15 11:39:38 INFO SchemaProvider: l -> string
20/12/15 11:39:38 INFO SchemaProvider: o -> string
20/12/15 11:39:38 INFO SchemaProvider: Retrieving schema for query: SELECT * FROM `http://dig.isi.edu/hasParent`
20/12/15 11:39:38 WARN SchemaProvider: Using ugly hack for adding a limit
20/12/15 11:39:38 INFO SchemaProvider: s -> string
20/12/15 11:39:38 INFO SchemaProvider: o -> string
20/12/15 11:39:38 INFO SchemaProvider: Retrieving schema for query: SELECT * FROM `http://www.w3.org/1999/02/22-rdf-syntax-ns#type`
20/12/15 11:39:38 WARN SchemaProvider: Using ugly hack for adding a limit
20/12/15 11:39:38 INFO SchemaProvider: s -> string
20/12/15 11:39:38 INFO SchemaProvider: o -> string
20/12/15 11:39:38 INFO SchemaProvider: Retrieving schema for query: SELECT * FROM `http://dig.isi.edu/age_XMLSchema#integer`
20/12/15 11:39:38 WARN SchemaProvider: Using ugly hack for adding a limit
20/12/15 11:39:38 INFO SchemaProvider: s -> string
20/12/15 11:39:38 INFO SchemaProvider: o -> string
20/12/15 11:39:38 INFO SchemaProvider: Retrieving schema for query: SELECT * FROM `http://dig.isi.edu/hasSpouse`
20/12/15 11:39:38 WARN SchemaProvider: Using ugly hack for adding a limit
20/12/15 11:39:38 INFO SchemaProvider: s -> string
20/12/15 11:39:38 INFO SchemaProvider: o -> string
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [0] Rewrite started.
20/12/15 11:39:38 WARN CandidateViewSelectorBase: JENA'S ALGEBRA OPTIMIZATION DISABLED
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [84] Candidate selection completed
20/12/15 11:39:38 WARN MappingOpsImpl: Single member union - should be avoided
20/12/15 11:39:38 WARN MappingOpsImpl: Single member union - should be avoided
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [101] Mapping rewrite completed
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [102] Sql translation completed
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [138] Sql optimization completed
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: [138] Done (excluding serialization)
20/12/15 11:39:38 INFO SparqlSqlStringRewriterImpl: Variable Definitions:
seed: (http://aksw.org/sparqlify/rdfTerm 1 ?s "" "") [[URI]]
seed__down_hasSpouse__down_name: (http://aksw.org/sparqlify/rdfTerm 2 ?h_1 ?l "") [[LITERAL]]

+----+----+--------------------------+
|l   |h_1 |s                         |
+----+----+--------------------------+
|    |John|http://dig.isi.edu/Mary   |
|    |Mary|http://dig.isi.edu/John   |
|null|null|http://dig.isi.edu/John_jr|
+----+----+--------------------------+
```

**Comment by pj001** [comment]
Hello, how do you learn the example

**Comment by LorenzBuehmann** [comment]
Should be fixed with commit 5a63496b88d5eb9dd5eb57f68a8a53126f031d5b

@carstendraschner can you verify this please?

**Comment by carstendraschner** [comment]
Thank you very much for your help! I the first samples I tried, It seems, that the issue is resolved :)
